File: java/sdk/sdk-core/src/main/java/com/novax/sdk/core/http/InterceptorChain.java
package com.novax.sdk.core.http;

import java.io.IOException;
import java.util.List;

/** Drives the interceptor chain and hands the final request to the transport. */
public final class InterceptorChain {

    private final HttpTransport transport;
    private final List<Interceptor> interceptors;

    public InterceptorChain(HttpTransport transport, List<Interceptor> interceptors) {
        this.transport = transport;
        this.interceptors = List.copyOf(interceptors);
    }

    public SdkResponse proceed(SdkRequest request) throws IOException {
        return new Node(request, 0).proceed(request);
    }

    private final class Node implements Interceptor.Chain {

        private final SdkRequest request;
        private final int index;

        Node(SdkRequest request, int index) {
            this.request = request;
            this.index = index;
        }

        @Override
        public SdkRequest request() {
            return request;
        }

        @Override
        public SdkResponse proceed(SdkRequest req) throws IOException {
            if (index < interceptors.size()) {
                return interceptors.get(index).intercept(new Node(req, index + 1));
            }
            return transport.execute(req);
        }
    }
}

File: python/novax_sdk/http/interceptor_chain.py
from __future__ import annotations

from .interceptor import Chain, Interceptor
from .sdk_request import SdkRequest
from .sdk_response import SdkResponse
from .transport import HttpTransport


class InterceptorChain:
    def __init__(self, transport: HttpTransport, interceptors: list[Interceptor]) -> None:
        self._transport = transport
        self._interceptors = list(interceptors)

    def proceed(self, request: SdkRequest) -> SdkResponse:
        return _Node(request, 0, self._interceptors, self._transport).proceed(request)


class _Node(Chain):
    def __init__(self, req: SdkRequest, index: int,
                 interceptors: list[Interceptor], transport: HttpTransport) -> None:
        self._req = req
        self._index = index
        self._interceptors = interceptors
        self._transport = transport

    @property
    def request(self) -> SdkRequest:
        return self._req

    def proceed(self, request: SdkRequest) -> SdkResponse:
        if self._index < len(self._interceptors):
            return self._interceptors[self._index].intercept(
                _Node(request, self._index + 1, self._interceptors, self._transport)
            )
        return self._transport.execute(request)
